pydantic/schema.py

```diff
@@ -946,6 +950,41 @@ def constraint_func(**kwargs: Any) -> Type[Any]:
 
     ans = go(annotation)
 
+    check_unused_constraints(constraints, used_constraints, field_name)
+
+    return ans
+
+
+def check_unused_assignment_constraints(field_info: FieldInfo, field_name: str, validate_assignment: bool) -> None:
```

Comment:
Maybe I'm missing something, but I think this is no longer required?

Reply:
It's still used here https://github.com/samuelcolvin/pydantic/pull/2196/files#diff-bfaaac2ae0a71d9686a272982d29d0bed82bfaacf8b5b30e2b88ed2543e74a3bR321

It's needed because `allow_mutation` requires a check of `validate_assignment`, none of the other constraints need that information.  If there is a different approach to adding this check which you would prefer, I'm open to ideas.  It might make sense to pull the `check_unused_constraints` call out `get_annotation_from_field_info`.  Having that constraint check hidden in there might be part of the confusion.

Reply:
I originally had a check within `get_annotation_from_field_info`, but `make` complained it made the function too complex.  Perhaps the function needs to be broken up instead of the approach I took.

Reply:
Took another stab at refactoring the `schema.get_annotation_from_field_info` function.   Basically just pulled out the check for unused constraints to be checked outside of the methods.